libs/contrib/Data/Fun/Extra.idr

```diff
@@ -0,0 +1,54 @@
+module Data.Fun.Extra
+
+import Data.Fun
+import Data.HVect
+
+||| Apply an n-ary function to an n-ary tuple of inputs
+public export
+apply : {ts : Vect n Type} -> {cod : Type} -> Fun ts cod -> HVect ts -> cod
+apply f [] = f
+apply f (t::ts) = apply (f t) ts
+
+||| Apply an n-ary function to an n-ary tuple of inputs
+public export
+applyPartially : {ts : Vect n Type} -> {ss : Vect m Type} -> {rs : Vect k Type} 
```

Comment:
This is the composition of `curry` and `Fun` seen as a homomorphism from
the monoid `HVect` to `Type -> Type`. Probably worth implementing the homo
parts independently:

```idris
homoNeutral : Fun [] cod -> id cod
homoMult    : Fun (rs ++ ss) cod -> (Fun rs . Fun ss) cod
```

Reply:
I see. I guess it's a special case of Cayley's theorem. Cool, I've added it. I've dropped the partial application from the dependent type, it's not really worth it. (It would be more interesting to generalise it to full telescopes.)